src/docfx/lib/collections/WorkQueue.cs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal class WorkQueue<T>
+    {
+        private readonly ConcurrentQueue<T> _queue = new ConcurrentQueue<T>();
+        private readonly ConcurrentHashSet<T> _recurseDetector = new ConcurrentHashSet<T>();
+
+        private readonly TaskCompletionSource<int> _drainTcs = new TaskCompletionSource<int>();
+
+        // For progress reporting
+        private int _totalCount = 0;
+        private int _processedCount = 0;
+
+        // For completion detection
+        private int _remainingCount = 0;
+
+        public void Enqueue(IEnumerable<T> items)
+        {
+            foreach (var item in items)
+            {
+                Enqueue(item);
+            }
+        }
+
+        public void Enqueue(T item)
+        {
+            if (_drainTcs.Task.IsCompleted)
+            {
+                throw new InvalidOperationException();
+            }
+
+            if (!_recurseDetector.TryAdd(item))
+            {
+                return;
+            }
+
+            _queue.Enqueue(item);
+
+            Interlocked.Increment(ref _totalCount);
+            Interlocked.Increment(ref _remainingCount);
+        }
+
+        public Task Drain(Func<T, Task> run, Action<int, int> progress = null)
```

Comment:
>Func<T, Task> run [](start = 26, length = 17)

may we make the `run` as a member of WorkQueue? then EnQueue and Drain will both use it.  Currently it's a little confuse for me that the relationship between Enqueue and Drain, seems Enqueue is using Drain parameter

Reply:
`Enqueue` will not use `run`, it just add it to the queue, this is intentional because we only have enough information to call `run` when `Drain` is called.

Reply:
What i mean is that Enqueue is using Drain parameter which is confusing me, but since it's limited by the existing workflow, i am ok with that.

---
In reply to: [294582612](https://github.com/dotnet/docfx/pull/4742#discussion_r294582612) [](ancestors = 294582612)